memstore/list/vector_party.go

```diff
@@ -0,0 +1,562 @@
+type baseVectorParty struct {
+	// offset is a pair of uint32 [offset, length]. therefore its length is 2 * length of vp.
+	offsets *memstore.Vector
+	// storing the offset to slab footer offset for each row.
+	length int
+	// DataType of values. We need it since for mode 0 vector party, we cannot
+	// get data type from values vector. Also we store it on disk anyway.
+	dataType common.DataType
+	// total number of elements.
+	totalElements int64
+	reporter      HostMemoryChangeReporter
+}
+
+// GetOffsetLength returns the <offset, length> pair at ith row.
+func (vp *baseVectorParty) GetOffsetLength(row int) (offset uint32, length uint32) {
+	offset = *(*uint32)(vp.offsets.GetValue(2 * row))
+	length = *(*uint32)(vp.offsets.GetValue(2*row + 1))
+	return
+}
+
+// GetValidity get validity of given offset. For list type it should always be true even for
+// empty list.
+func (vp *baseVectorParty) GetValidity(offset int) bool {
+	return true
+}
+
+// GetDataValue just return null data value for list type vp. It's questionable to store list element value in
+// this common.DataValue since this struct is already too huge.
+// TODO(lucafuji): figure out how to read individual element from list vector party.
+func (vp *baseVectorParty) GetDataValue(offset int) common.DataValue {
+	return common.NullDataValue
+}
+
+// SetDataValue is not supported by list vector party.
+func (vp *baseVectorParty) SetDataValue(offset int, value common.DataValue,
+	countsUpdateMode common.ValueCountsUpdateMode, counts ...uint32) {
+	utils.GetLogger().Panic("SetDataValue is not supported by list vector party")
+}
+
+// GetDataValueByRow just returns a null data value for now.
+// TODO(lucafuji): figure out how to read individual element from list vector party.
+func (vp *baseVectorParty) GetDataValueByRow(row int) common.DataValue {
+	return common.NullDataValue
+}
+
+// IsList tells whether this vp is list vp. And user can later on cast it to proper interface.
+func (vp *baseVectorParty) IsList() bool {
+	return true
+}
+
+// GetDataType returns the element date type of this vp.
+func (vp *baseVectorParty) GetDataType() common.DataType {
+	return vp.dataType
+}
+
+// GetLength returns the length of the vp.
+func (vp *baseVectorParty) GetLength() int {
+	return vp.length
+}
+
+// GetElementLength returns the length of ith element
+func (vp *baseVectorParty) GetElementLength(row int) int {
+	_, l := vp.GetOffsetLength(row)
+	return int(l)
+}
+
+// Slice vector party into human readable SlicedVector format. For now just return an
+// empty slice.
+// TODO(lucafuji): implement slice vector on list vp.
+func (vp *baseVectorParty) Slice(startRow, numRows int) common.SlicedVector {
+	return common.SlicedVector{}
+}
+
+// Check whether two vector parties are equal (used only in unit tests)
+// Check common properties like data type and length.`
+func (vp *baseVectorParty) equals(other common.VectorParty, listVP common.ListVectorParty) bool {
```

Comment:
this equals is lower case, so it's an unexported help function called by ArchiveListVP and LivingListVP's Equals to avoid code duplication. See line 221 and line 422